### PR TITLE
[SPARK-45509][SQL] Fix df column reference behavior for Spark Connect

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This PR fixes a few problems of column resolution for Spark Connect, to make the behavior closer to classic Spark SQL (unfortunately we still have some behavior differences in corner cases).
1. resolve df column references in both `resolveExpressionByPlanChildren` and `resolveExpressionByPlanOutput`. Previously it's only in `resolveExpressionByPlanChildren`.
2. when the plan id has multiple matches, fail with `AMBIGUOUS_COLUMN_REFERENCE`

### Why are the changes needed?

fix behavior differences between spark connect and classic spark sql

### Does this PR introduce _any_ user-facing change?

Yes, for spark connect scala client

### How was this patch tested?

new tests

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 1]
```diff
@@ -872,6 +872,68 @@ class ClientE2ETestSuite extends RemoteSparkSession with SQLHelper with PrivateM
...
+
+    checkSameResult(
+      Seq(Row(1, "a")),
+      // All these column references are not ambiguous and are still valid after join.
+      df1.join(df2, df1("i") + 1 === df2("i")).sort(df1("i").desc).select(df1("i"), df1("j"))
+    )
+
+    val e2 = intercept[AnalysisException] {
+      // df1("i") is ambiguous as df1 appears in both join sides.
+      df1.join(df1, df1("i") === 1).collect()
```
Comment (cloud-fan) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:902]: classic spark sql thinks this is not ambiguous. It's probably a bug and I'll fix later.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 2]
```diff
@@ -872,6 +872,68 @@ class ClientE2ETestSuite extends RemoteSparkSession with SQLHelper with PrivateM
...
+    val e4 = intercept[AnalysisException] {
+      // df1("i") is ambiguous as df1 appears in both join sides (df1_filter contains df1).
+      df1.join(df1_filter, df1("i") === 1).collect()
+    }
+    assert(e4.getMessage.contains("AMBIGUOUS_COLUMN_REFERENCE"))
+
+    checkSameResult(
+      Seq(Row("a")),
+      // df1_filter("i") is not ambiguous as df1_filter does not exist in the join left side.
+      df1.join(df1_filter, df1_filter("i") === 1).select(df1_filter("j"))
```
Comment (cloud-fan) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:921]: Classic Spark SQL thinks this is ambiguous, as it uses `AttributeReference` directly and we are not able to re-resolve it. Spark Connect uses `UnresolvedAttribute` which is lazy binding and works fine in this case.

**Comment by cloud-fan** [comment]
cc @zhengruifeng @hvanhovell @HyukjinKwon 

I think the Spark Connect behavior is very reasonable now. We should move Classic Spark SQL to this behavior as well in the future.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala [hunk 3]
```diff
@@ -539,4 +533,28 @@ trait ColumnResolutionHelper extends Logging {
         None
     }
   }
+
+  private def findPlanById(
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala:537]: can we do tailrec?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala [hunk 4]
```diff
@@ -490,39 +477,46 @@ trait ColumnResolutionHelper extends Logging {
     }
   }
 
-  private def resolveExpressionByPlanId(
+  // If the TreeNodeTag 'LogicalPlan.PLAN_ID_TAG' is attached, it means that the plan and
+  // expression are from Spark Connect, and need to be resolved in this way:
+  //    1. extract the attached plan id from UnresolvedAttribute;
+  //    2. top-down traverse the query plan to find the plan node that matches the plan id;
+  //    3. if can not find the matching node, fail the analysis due to illegal references;
+  //    4. if more than one matching nodes are found, fail due to ambiguous column reference;
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala:485]: Just for my own understanding, does it mean that it fails when more than one matching nodes are found within the same level children?

**Review by HyukjinKwon — APPROVED** [review]

File: common/utils/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -31,6 +31,15 @@
     ],
     "sqlState" : "42702"
   },
+  "AMBIGUOUS_COLUMN_REFERENCE" : {
+    "message" : [
+      "Column <name> is ambiguous. It's because you joined several DataFrame together, and some of these DataFrames are the same.",
```
Comment (zhengruifeng) [inline common/utils/src/main/resources/error/error-classes.json:36]: Can we reuse `AMBIGUOUS_REFERENCE (42704)`?

and I guess another cause maybe dataframe creation with duplicated column names:
```
scala> val df = Seq((1,2), (3,4)).toDF("a","a")
val df: org.apache.spark.sql.DataFrame = [a: int, a: int]
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala:485]: within the sub-plan under the plan node that holds the column reference

File: common/utils/src/main/resources/error/error-classes.json (hunk 5 above)
Comment (cloud-fan) [reply common/utils/src/main/resources/error/error-classes.json:36]: It's different. This is for found more than one matching dataframes, but `AMBIGUOUS_REFERENCE` is for found more than one matching attribute from candidates.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala:537]: can't find a way to make it tailrec... I need to check the return value from `plan.children.flatMap(findPlanById(u, id, _))`

File: common/utils/src/main/resources/error/error-classes.json (hunk 5 above)
Comment (zhengruifeng) [reply common/utils/src/main/resources/error/error-classes.json:36]: sounds good.

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/sql/connect/plan.py [hunk 6]
```diff
@@ -2214,7 +2214,9 @@ def __init__(
         self._input_grouping_cols = input_grouping_cols
         self._other_grouping_cols = other_grouping_cols
         self._other = cast(LogicalPlan, other)
-        self._func = function._build_common_inline_user_defined_function(*cols)
+        # The function takes entire DataFrame as inputs, no need to do
+        # column binding (no input columns).
+        self._func = function._build_common_inline_user_defined_function()
```
Comment (cloud-fan) [inline python/pyspark/sql/connect/plan.py:2219]: cc @zhengruifeng @HyukjinKwon

Comment (zhengruifeng) [reply python/pyspark/sql/connect/plan.py:2219]: it seems this field is not used in connect

https://github.com/apache/spark/blob/48e207f4a2192d474f2c0f141b984ef0c36a78c3/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala#L694-L723

cc @xinrong-meng

Comment (cloud-fan) [reply python/pyspark/sql/connect/plan.py:2219]: yea, that's why this bug is hidden. We simply picked the first matching dataframe in the query plan to resolve columns, which is wrong and now we will throw AMBIGUOUS_COLUMN_REFERENCE

**Review by dongjoon-hyun — COMMENTED** [review]
Could you re-trigger the failed `catalyst` test pipeline?

**Comment by cloud-fan** [comment]
thanks for the reviews, merging to master/3.5!

**Comment by cloud-fan** [comment]
The 3.5 backport PR: https://github.com/apache/spark/pull/43699

**Pull request closed without merging**